### PR TITLE
Add softmax_focal_loss() to allow multi-class focal loss

### PR DESCRIPTION
**Pull request by dhruvbird**
In image segmentation tasks, focal loss is useful when trying to classify an image pixel as one of N classes. Unfortunately, sigmoid_focal_loss() isn't useful in such cases. I found that other have been asking for this as well here https://github.com/pytorch/vision/issues/3250 so I decided to submit a PR for the same.

I'm opening this PR to check if this is something the pytorch-vision team is interested in merging.

### CODE REVIEW

File: torchvision/ops/focal_loss.py [hunk 1]
```diff
@@ -56,3 +56,81 @@ def sigmoid_focal_loss(
...
+                The predictions for each example. Softmax() is applied on this tensor
+                to convert the raw logits to class probabilities. Expected shape is
+                (N, C, *).
+        targets (Tensor): Must be a long tensor similar to the one expected by
+                PyTorch's CrossEntropyLoss.
+                https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html
+                The class dimension is expected to be absent, and each
+                element is the class value in the range [0, C).
+        alpha (float): Weighting factor in range (0,1) to balance
+                positive vs negative examples or -1 for ignore. Default: ``0.25``.
```
Comment (rehno-lindeque) [inline torchvision/ops/focal_loss.py:81]: I don't see a branch checking if alpha is `-1` anywhere.

File: torchvision/ops/focal_loss.py [hunk 2]
```diff
@@ -56,3 +56,81 @@ def sigmoid_focal_loss(
...
+    #
+    # Hence, exp(-loss) == pt
+    #
+    # This trick works only if the targets is a long tensor. If it's a float
+    # tensor, then each each value is a probability, and we'd need to divide
+    # the result of cross entropy loss by the probability, and hence would
+    # need to compute the softmax manually anyway. We don't implement that
+    # here for brevity, but this code can be extended for such a use-case.
+    pt = torch.exp(-ce_loss)
+    focal_loss = alpha * ((1 - pt) ** gamma) * ce_loss
```
Comment (rehno-lindeque) [inline torchvision/ops/focal_loss.py:124]: I have my doubts that the `alpha` term used here is correct. In fact we want `alpha_t`.

Although the paper doesn't explicitly give the formula for `alpha_t`, it states in it's definition

> For notational convenience, we define $`α_t$` analogously to how we defined $`p_t$`

So given that

$`
\begin{align*}
p_t =
\begin{cases}
\alpha & \text{if } y = 1 \\
(1 - p) & \text{otherwise}
\end{cases}
\end{align*}`$

I believe $`\alpha_t` should be analogously interpreted as 

$`
\begin{align*}
\alpha_t =
\begin{cases}
\alpha & \text{if } y = 1 \\
(1 - \alpha) & \text{otherwise}
\end{cases}
\end{align*}`$

This seems to be the case in torchvision

https://github.com/pytorch/vision/blob/cc0f9d02e1e7549e9e8ef1081a0d3b5d72f1297d/torchvision/ops/focal_loss.py#L43-L45

I've seen another implementation in the wild pass through `alpha` to the `weights` of `nll_loss`: 

[AdeelH/pytorch-multi-class-focal-loss/focal_loss.py](https://github.com/AdeelH/pytorch-multi-class-focal-loss/blob/2477bf68ac61d2d2e5e044c63f70b156073c5ffd/focal_loss.py#L48-L49)

This strikes me as the correct approach since it allows one to weigh each class separately in a multi-class setting.

Comment (dhruvbird) [reply torchvision/ops/focal_loss.py]: I agree with your assessment above.

I have put up a revised PR at https://github.com/pytorch/vision/pull/7760 since I switched local branches.
